vopono_core/src/network/piapf.rs

```diff
@@ -0,0 +1,193 @@
+pub struct Piapf {
+    pub port: u16,
+    loop_thread_handle: Option<JoinHandle<()>>,
+    send_channel: Sender<bool>,
+}
+
+struct ThreadParams {
+    pub port: u16,
+    pub netns_name: String,
+    pub signature: String, 
+    pub payload: String,
+    pub hostname: String,
+    pub gateway: String,
+}
+
+impl Piapf {
+    pub fn new(ns: &NetworkNamespace, protocol: &Protocol) -> anyhow::Result<Self> {
+        let pia = PrivateInternetAccess {}; //This is a bit weird, no? There's no state, so effectively all the methods are static...
+        
+        let traceroute_response = NetworkNamespace::exec_with_output(
+            &ns.name, 
+            &["traceroute", "-n", "-m", "1", "privateinternetaccess.com" ], )?;
+        if !traceroute_response.status.success() {
+            log::error!("Could not locate gateway with traceroute");
+            anyhow::bail!("Could not locate gateway with traceroute")
+        }
+        let re = Regex::new(r" *1 *(?P<gateway>\d{1,3}.\d{1,3}.\d{1,3}.\d{1,3}).*").unwrap();
+        let result = String::from_utf8_lossy(&traceroute_response.stdout);
+        let second_line = result.lines().skip(1).next().unwrap();
+        let vpn_gateway = re.captures(&second_line).unwrap().get(1).unwrap().as_str().to_string();
+        
+        log::info!("PIA gateway: {}", vpn_gateway);
+        
+        let vpn_hostname = match protocol {
+            Protocol::OpenVpn => "nl-amsterdam.privacy.network".to_string(), // FIXME: Parse this from the OpenVPN conf?
+            Protocol::Wireguard => "nl-amsterdam.privacy.network".to_string(), // FIXME: [Insert clever idea to get wireguard endpoint hostname here] 
+            _ => {
+                log::error!("PIA port forwarding only supported for OpenVPN and Wireguard");
+                anyhow::bail!("PIA port forwarding only supported for OpenVPN and Wireguard")
+            }
+        };
+        
+        log::info!("PIA hostname: {}", vpn_hostname);
+        
+        let (pia_user, pia_pass) = match protocol {
+            Protocol::OpenVpn => pia.load_openvpn_auth()?,
+            Protocol::Wireguard => pia.load_wireguard_auth()?,
+            _ => {
+                log::error!("PIA port forwarding only supported for OpenVPN and Wireguard");
+                anyhow::bail!("PIA port forwarding only supported for OpenVPN and Wireguard")
+            }
+        };
+        
+        //log::info!("PIA u/p: {} / {}", pia_user, pia_pass);
+        
+        let pia_token = PrivateInternetAccess::get_pia_token(&pia_user, &pia_pass)?;
+        
+        log::info!("PIA pia_token: {}", pia_token);
+
+        let get_response = NetworkNamespace::exec_with_output(&ns.name, &["curl", 
+            "-s", "-m", "5",
+            "--connect-to", &format!("{}::{}:", vpn_hostname, vpn_gateway).to_string(),
+            "--cacert", "/home/benland100/vopono/vopono_core/src/config/providers/pia/ca.rsa.4096.crt", //FIXME: how to get this path?
```

Comment:
We should download the file to the XDG config vopono directory - i.e. `~/.config/vopono/pia/`

Reply:
See below re. it already being embedded in the binary, we should write that out at sync time (or move it to be a public part of the PIA struct).